internal/agent/agent.go

```diff
@@ -0,0 +1,251 @@
+}
+func (m *MetricsCollector) AdditionalMetrics() {
+	v, _ := mem.VirtualMemory()
+	cpu, _ := cpu.Percent(time.Second, true)
+	m.mx.Lock()
+	defer m.mx.Unlock()
+	m.GaugeMetrics["TotalMemory"] = v.Total
+	m.GaugeMetrics["FreeMemory"] = v.Free
+	for i, unit := range cpu {
+		key := "CPUutilization" + strconv.Itoa(i)
+		m.GaugeMetrics[key] = unit
+	}
+}
+func (m *MetricsCollector) GetData() []metric.Metrics {
+	op := "agent.GetData"
+	m.mx.Lock()
+	defer m.mx.Unlock()
+	var metrics []metric.Metrics
+	for key, val := range m.GaugeMetrics {
+		typedValue, err := convertToFloat64(val)
+		if err != nil {
+			logger.Log.Debug("failed to Convert GaugeMetrics value",
+				zap.String("operation", op),
+				zap.String("value", fmt.Sprintf("%v", val)),
+			)
+		}
+
+		metric := metric.Metrics{
+			ID:    key,
+			MType: metric.MetricTypeGauge,
+			Value: &typedValue,
+		}
+		metrics = append(metrics, metric)
+	}
+	value := new(int64)
+	*value = m.PollCount
+	metric := metric.Metrics{
+		ID:    "PollCount",
+		MType: metric.MetricTypeCounter,
+		Delta: value,
+	}
+
+	metrics = append(metrics, metric)
+	return metrics
+}
+func (m *MetricsCollector) Send(url string) error {
+	op := "agent.Send"
+	_ = op
+	metrics := m.GetData()
+	if m.cfg.RateLimit > 0 {
+		go func() {
+			for _, metric := range metrics {
+				m.jobs <- metric
+			}
+		}()
+	} else {
+		_, err := m.sendMetric(url, metrics)
+
+		if err != nil {
+			return err
+		}
+	}
+	m.mx.Lock()
+	m.PollCount = 0
+	m.mx.Unlock()
+	return nil
+}
+func (m *MetricsCollector) sendMetric(url string, metricObj []metric.Metrics) (int, error) {
+	op := "agent.SendMetric"
+	requestBody, err := json.Marshal(metricObj)
+	if err != nil {
+		log.Print(err)
+	}
+
+	client := &http.Client{}
+	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(requestBody))
+	if err != nil {
+		log.Print(err)
+	}
+	if len(m.cfg.SignBodyKey) > 0 {
+		header := hmac.SignData(requestBody, m.cfg.SignBodyKey)
+		req.Header.Add(SignatureHeaderName, header)
+	}
+	for i := 0; i < retryCount; i++ {
+		res, err := client.Do(req)
+
+		if err == nil {
+			res.Body.Close()
+			return res.StatusCode, nil
+		}
+
+		logger.Log.Error("can't sent metrics",
+			zap.String("operation", op),
+			zap.String("trying to resend metrics:", fmt.Sprintf("tries number- %d", i+1)),
+			zap.Error(err),
+		)
+		time.Sleep(time.Duration(retryInterval[i]) * time.Second)
+	}
+	return http.StatusInternalServerError, nil
+}
+func (m *MetricsCollector) Run() {
+	op := "agent.Run"
+	url := "http://" + m.cfg.ServerAddr + "/updates/"
+	tickerUpdateMetrics := time.NewTicker(time.Duration(m.cfg.PollInterval) * time.Second)
+	quitUpdateMetrics := make(chan struct{})
+	if m.cfg.RateLimit > 0 {
+		m.jobs = make(chan metric.Metrics, NumJobs)
+		for w := 1; w <= m.cfg.RateLimit; w++ {
+			go m.worker(w, url)
+		}
+	}
+	go func() {
+		for {
+			select {
+			case <-tickerUpdateMetrics.C:
+				m.UpdateMetrics()
+				go m.AdditionalMetrics()
```

Comment:
```suggestion
				wg := &sync.WaitGroup
				wg.Add(1)
				go m.AdditionalMetrics(wg)
				m.UpdateMetrics()
				wg.Wait()
```
Внутри AdditionalMetrics необходимо обязательно дернуть в конце (можно через defer) wg.Done(). В этом случае мы сможем синхронизировать окончание сбора метрик между двумя горутинами. И на сервер уедет чуть более консистентное состояние.